### PR TITLE
[BugFix] fix orc pushdown filter failure

### PR DESCRIPTION
**Pull request by kobebryantlin0**
## What type of PR is this：
BugFix
 
## Problem Summary(Required) ：
When ORC executes filter pushdown, assuming there are two predicates in the SQL query, when the function `filterOnPickStringDictionary` is executed, if predicate1 is present in `sdicts`, dictionary filtering can be performed and `dict_value_chunk->num_rows()` after filtering is greater than 0, then `_dict_filter_eval_cache` will not be cleared. However, if predicate2 is not present in `sdicts`, the dictionary filtering in `filterOnPickStringDictionary` will not be executed, and `eval_conjuncts_into_filter` will obtain `chunk_size` of 0, which will execute the next chunk retrieval operation. It can be seen that `_dict_filter_eval_cache` will not be properly cleared. When the next batch of stripe data is retrieved and `_orc_row_reader_filter->is_slot_evaluated(it.first)` returns true, it will skip verification even if it does not satisfy the predicate."

### CODE REVIEW

**Comment by kobebryantlin0** [comment]
@Smith-Cruise

**Comment by dirtysalt** [comment]
Thanks @kobebryantlin0 . It's quite hard to catch that.

Could you be kind to give an exmaple/ut to reproduce this case?

**Comment by kobebryantlin0** [comment]
> Thanks @kobebryantlin0 . It's quite hard to catch that.
> 
> Could you be kind to give an exmaple/ut to reproduce this case?

**Example**

`SELECT` id, `name` FROM table WHERE dt >= CAST("2023-03-30" AS string) AND dt <= CAST("2023-03-30" AS string) AND id = '003' AND `outline_1` = '投诉' limit `1;`

+-----------------+ 
| id.   | name |
+-----------------+
| 001 | 咨询 |
| 002 | 投诉 |

| 003 | 反馈 | 
| 004 | 咨询 |
+-----------------+-

**Real Output:**

+-----------------+ 
| id | name |
+-----------------+

| 003 | 反馈 |

**Expected Output:**

empty set


In the sample data above, 001 and 002 belong to the first batch of chunks, while 003 and 004 belong to the second batch of chunks. In the first batch of two data rows, the `id` field returns false for `filterOnPickStringDictionary` because `sdicts` is empty, so data filtering is not performed. For the `002` record, the `name` field matches, so `_dict_filter_eval_cache` is not cleared. Moving on to the second batch of chunk data, the `id` field matches for the `003` record, but the `name` field is not validated because the `_dict_filter_eval_cache` from the previous batch was not cleared, so data filtering does not take effect in this case.

**Comment by DorianZheng** [comment]
please add ut

**Comment by kobebryantlin0** [comment]
> Does it mean t
> 
> > > Thanks @kobebryantlin0 . It's quite hard to catch that.
> > > Could you be kind to give an exmaple/ut to reproduce this case?
> > 
> > 
> > **Example**
> > **Raw Data** +-----------------+ | id. | name | +-----------------+ | 001 | 咨询 | | 002 | 投诉 |
> > | 003 | 反馈 | | 004 | 咨询 | +-----------------+-
> > **Query Sql**
> > `SELECT` id, `name` FROM table WHERE dt >= CAST("2023-03-30" AS string) AND dt <= CAST("2023-03-30" AS string) AND id = '003' AND `name` = '投诉' limit `1;`
> > **Real Output:**
> > +-----------------+ | id | name | +-----------------+
> > | 003 | 反馈 |
> > **Expected Output:**
> > empty set
> > In the sample data above, 001 and 002 belong to the first batch of chunks, while 003 and 004 belong to the second batch of chunks. In the first batch of two data rows, the `id` field returns false for `filterOnPickStringDictionary` because `sdicts` is empty, so data filtering is not performed. For the `002` record, the `name` field matches, so `_dict_filter_eval_cache` is not cleared. Moving on to the second batch of chunk data, the `id` field matches for the `003` record, but the `name` field is not validated because the `_dict_filter_eval_cache` from the previous batch was not cleared, so data filtering does not take effect in this case.
> > @Smith-Cruise @dirtysalt
> 
> Good catch!
> 
> In this example, do you mean there are two stripe: stripe one contains two row: | 001 | 咨询 | | 002 | 投诉 |
> 
> stripe two contains other two row: | 003 | 反馈 | | 004 | 咨询 |
> 
> For the first stripe, there exists `sdicts`, so the `_dict_filter_eval_cache` is evaluated. but for the second stripe, there are no `sdicts`, so the `_dict_filter_eval_cache` is not evaluated and not cleaned. later we use the wrong `_dict_filter_eval_cache` to filter data
> 
> as you can see, if `sdicts` is empty, it means `StringVectorBatch::use_codes` will be false too, so there is another issue is in `OrcChunkReader::apply_dict_filter_eval_cache` we don't check if `StringVectorBatch::use_codes` is false or not to decide whether to filter it by `_dict_filter_eval_cache`

 @DorianZheng 

You are right, there was probably an issue before, but with the changes I made in my submitted PR, I believe we do not need to check for it in `apply_dict_filter_eval_cache` anymore. This is because after my modifications, `filterOnPickStringDictionary` always clear `dict_filter_eval_cache` before returning, regardless of whether sdicts is empty or not. Since `filterOnPickStringDictionary` is executed before `apply_dict_filter_eval_cache`, theoretically `dict_filter_eval_cache` should have already been cleared in the former if use_codes is false. Additionally, `apply_dict_filter_eval_cache` has a check for whether `dict_filter_eval_cache.size() `is 0, so subsequent logic should not be executed.

**Comment by DorianZheng** [comment]
@kobebryantlin0 Thanks for the fix. Could you add UT.

**Comment by dirtysalt** [comment]
To normal orc writer, there will never be the case that two stripes use different dictionary encoding schema.  Because there is a configuration item said that

```
  ROW_INDEX_STRIDE_DICTIONARY_CHECK("orc.dictionary.early.check",
      "hive.orc.row.index.stride.dictionary.check",
      true,
      "If enabled dictionary check will happen after first row index stride\n" +
          "(default 10000 rows) else dictionary check will happen before\n" +
          "writing first stripe. In both cases, the decision to use\n" +
          "dictionary or not will be retained thereafter."),
```

So to reproduce this case, I have to make some modification on java orc code like following:

```java
diff --git a/java/core/src/java/org/apache/orc/impl/writer/StringBaseTreeWriter.java b/java/core/src/java/org/apache/orc/impl/writer/StringBaseTreeWriter.java
index 9d87873e7..d320df7b4 100644
--- a/java/core/src/java/org/apache/orc/impl/writer/StringBaseTreeWriter.java
+++ b/java/core/src/java/org/apache/orc/impl/writer/StringBaseTreeWriter.java
@@ -139,6 +139,7 @@ public abstract class StringBaseTreeWriter extends TreeWriterBase {
     // reset all of the fields to be ready for the next stripe.
     dictionary.clear();
     savedRowIndex.clear();
+    doneDictionaryCheck = false;
     rowIndexValueCount.clear();
     if (rowIndexPosition != null) {
       recordPosition(rowIndexPosition);
```

------


And by generating following file and table schema, I can reproduce this case

```java
package org.apache.orc.examples;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.hive.ql.exec.vector.BytesColumnVector;
import org.apache.hadoop.hive.ql.exec.vector.LongColumnVector;
import org.apache.hadoop.hive.ql.exec.vector.VectorizedRowBatch;
import org.apache.orc.OrcFile;
import org.apache.orc.TypeDescription;
import org.apache.orc.Writer;

import java.io.File;
import java.util.Random;

public class GenORCFileExample {
    public static String genRandomString(int size) {
        Random rnd = new Random();
        StringBuffer sb = new StringBuffer();
        for (int i = 0; i < size; i++) {
            char ch = (char) (rnd.nextInt(100) % 26 + 'a');
            sb.append(ch);
        }
        return sb.toString();
    }

    public static void genTwoStripsDictAndNoDict() throws Exception {
        Configuration conf = new Configuration();
        int batchSize = 100;
        conf.setDouble("orc.dictionary.key.threshold", 0.01);
        conf.setLong("orc.stripe.size", 10);
        conf.setLong("orc.stripe.row.count", batchSize);

        TypeDescription schema = TypeDescription.fromString("struct<x:int,y:string>");
        Path p = new Path("/Users/dirlt/Downloads/two-strips-dict-and-nodict.orc");
        File f = new File(p.toString());
        if (f.exists()) {
            f.delete();
        }
        Writer writer = OrcFile.createWriter(p,
                OrcFile.writerOptions(conf)
                        .setSchema(schema));

        VectorizedRowBatch batch = schema.createRowBatch(batchSize);
        LongColumnVector x = (LongColumnVector) batch.cols[0];
        BytesColumnVector y = (BytesColumnVector) batch.cols[1];
        for (int r = 0; r < batchSize; ++r) {
            x.vector[r] = r % 10;
            y.setVal(r, "a".getBytes());
        }
        batch.size = batchSize;
        writer.addRowBatch(batch);
        batch.reset();

        x.vector[0] = 1;
        y.setVal(0, "a".getBytes());
        for (int r = 1; r < batchSize; ++r) {
            x.vector[r] = r % 10;
            y.setVal(r, genRandomString(20).getBytes());
        }
        batch.size = batchSize;
        writer.addRowBatch(batch);
        batch.reset();

        writer.close();
    }

    public static void main(String[] args) {
        try {
            genTwoStripsDictAndNoDict();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

}
```

table schema

```
CREATE TABLE `orc_two_stripes_dict_and_nodict`(
  `x` int,
  `y` string)
ROW FORMAT SERDE
  'org.apache.hadoop.hive.ql.io.orc.OrcSerde'
STORED AS INPUTFORMAT
  'org.apache.hadoop.hive.ql.io.orc.OrcInputFormat'
OUTPUTFORMAT
  'org.apache.hadoop.hive.ql.io.orc.OrcOutputFormat'
LOCATION
  'hdfs://emr-header-1.cluster-49091:9000/user/hive/warehouse/zya.db/orc_two_stripes_dict_and_nodict'
TBLPROPERTIES (
  'transient_lastDdlTime'='1682014085')
```

And here is the wrong result

image

**Review by dirtysalt — DISMISSED** [review]

**Comment by dirtysalt** [comment]
I've verified that your pr works. I've approved that. 

But if you can add some UT, it would be better.

Can you take a look and see if you can add a unit test with this code? The unit test should load the 'hdfs_scanner_test.cpp' file, which contains several test cases for ORC (Optimized Row Columnar). I can send you the orc file I generated.

**Review by DorianZheng — DISMISSED** [review]

**Review by Smith-Cruise — DISMISSED** [review]

**Comment by imay** [comment]
@kobebryantlin0 Thank you for your contribution, could you clang-format your code and sign-off by `git commit -s`

**Review by wyb — DISMISSED** [review]

**Comment by kobebryantlin0** [comment]
> @kobebryantlin0 Thank you for your contribution, could you clang-format your code and sign-off by `git commit -s`

Sure, I can do that, but it may take a while as I'm currently at work.

**Comment by kobebryantlin0** [comment]
> @kobebryantlin0 Thanks for the fix. Could you add a unit test.

@DorianZheng @Smith-Cruise ut have been added.

**Review by Smith-Cruise — DISMISSED** [review]

**Review by Smith-Cruise — DISMISSED** [review]

File: be/test/exec/hdfs_scanner_test.cpp [hunk 1]
```diff
@@ -637,6 +637,100 @@ TEST_F(HdfsScannerTest, TestOrcGetNextWithDictFilter) {
...
+   Stream: column 1 section DATA start: 237 length 20
+   Stream: column 2 section DATA start: 257 length 1224
+   Stream: column 2 section LENGTH start: 1481 length 10
+   Encoding column 0: DIRECT
+   Encoding column 1: DIRECT_V2
+   Encoding column 2: DIRECT_V2
+*/
+
+TEST_F(HdfsScannerTest, TestOrcGetNextWithDiffEncodeDictFilter) {
+    SlotDesc two_diff_encode_orc_desc[] = {{"x", TypeDescriptor::from_logical_type(LogicalType::TYPE_BIGINT)},
```
Comment (dirtysalt) [inline be/test/exec/hdfs_scanner_test.cpp:683]: it's not BIGINT, it's INT.

the orc schema is

```
Type: struct<x:int,y:string>
```

File: be/test/exec/hdfs_scanner_test.cpp [hunk 2]
```diff
@@ -637,6 +637,100 @@ TEST_F(HdfsScannerTest, TestOrcGetNextWithDictFilter) {
...
+                                           {""}};
+    const std::string two_diff_encode_orc_file = "./be/test/exec/test_data/orc_scanner/two-strips-dict-and-nodict.orc";
+
+    auto scanner = std::make_shared<HdfsOrcScanner>();
+
+    auto* range = _create_scan_range(two_diff_encode_orc_file, 0, 0);
+    auto* tuple_desc = _create_tuple_desc(two_diff_encode_orc_desc);
+    auto* param = _create_param(two_diff_encode_orc_file, range, tuple_desc);
+
+    // key = "a", y = 'a' rows is 101
```
Comment (dirtysalt) [inline be/test/exec/hdfs_scanner_test.cpp:694]: the condition is x =1 and y = 'a'

Comment (dirtysalt) [reply be/test/exec/hdfs_scanner_test.cpp:694]: Oh, yes, it should reproduce the bug only with y='a' predicate.

File: be/test/exec/hdfs_scanner_test.cpp (hunk 1 above)
Comment (kobebryantlin0) [reply be/test/exec/hdfs_scanner_test.cpp]: @dirtysalt  Thank you for your cr and suggestion, i got it.  I also appreciate the help from @Smith-Cruise

**Review by Smith-Cruise — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Comment by kobebryantlin0** [comment]
> All checks have passed

**Comment by kobebryantlin0** [comment]
@DorianZheng @imay all checks have passed, could you help me merge pr into master? thx.

**Review by stephen-shelby — APPROVED** [review]

**Comment by Smith-Cruise** [comment]
@Mergify backport branch-3.0

**Pull request merged**